cpp/src/decisiontree/batched-levelalgo/kernels.cuh

```diff
@@ -403,35 +399,37 @@ __global__ void computeSplitRegressionKernel(
   }
   __threadfence();  // for commit guarantee
   __syncthreads();
-  // for MAE computation, we'd need a 2nd pass over data :(
+
+  /* Make a second pass over the data to compute gain */
+  // Wait until all blockIdx.x's are done
+  MLCommon::GridSync gs(workspace, MLCommon::SyncType::ACROSS_X, false);
+  gs.sync();
+  // now, compute the mean value to be used for metric update
+  for (IdxT i = threadIdx.x; i < nbins; i += blockDim.x) {
+    scount[i] = count[gcOffset + i];
```

Comment:
I agree. I had to read the code very carefully to deduce their meaning.

Reply:
Agree with John. When this was written initially, readability was not kept as a priority item! @hcho3 can you please file an issue so that we don't forget about this?